ecr-login/api/client.go

```diff
@@ -16,47 +16,70 @@ package api
 import (
 	"encoding/base64"
 	"fmt"
+	"net/url"
 	"regexp"
 	"strings"
 	"time"
 
-	"github.com/pkg/errors"
-
 	"github.com/aws/aws-sdk-go/aws"
 	"github.com/aws/aws-sdk-go/service/ecr"
-	"github.com/awslabs/amazon-ecr-credential-helper/ecr-login/cache"
+	"github.com/aws/aws-sdk-go/service/ecrpublic"
+	"github.com/pkg/errors"
 	"github.com/sirupsen/logrus"
+
+	"github.com/awslabs/amazon-ecr-credential-helper/ecr-login/cache"
+)
+
+const (
+	proxyEndpointScheme = "https://"
+	programName         = "docker-credential-ecr-login"
+	ecrPublicName       = "public.ecr.aws"
+	ecrPublicEndpoint   = proxyEndpointScheme + ecrPublicName
 )
 
-const proxyEndpointScheme = "https://"
-const programName = "docker-credential-ecr-login"
+var ecrPattern = regexp.MustCompile(`(^[a-zA-Z0-9][a-zA-Z0-9-_]*)\.dkr\.ecr(\-fips)?\.([a-zA-Z0-9][a-zA-Z0-9-_]*)\.amazonaws\.com(\.cn)?$`)
+
+type Service string
 
-var ecrPattern = regexp.MustCompile(`(^[a-zA-Z0-9][a-zA-Z0-9-_]*)\.dkr\.ecr(\-fips)?\.([a-zA-Z0-9][a-zA-Z0-9-_]*)\.amazonaws\.com(\.cn)?`)
+const (
+	ServiceECR       Service = "ecr"
+	ServiceECRPublic Service = "ecr-public"
+)
 
 // Registry in ECR
 type Registry struct {
-	ID     string
-	FIPS   bool
-	Region string
+	Service Service
+	ID      string
+	FIPS    bool
+	Region  string
 }
 
 // ExtractRegistry returns the ECR registry behind a given service endpoint
-func ExtractRegistry(serverURL string) (*Registry, error) {
-	if strings.HasPrefix(serverURL, proxyEndpointScheme) {
-		serverURL = strings.TrimPrefix(serverURL, proxyEndpointScheme)
+func ExtractRegistry(input string) (*Registry, error) {
+	if strings.HasPrefix(input, proxyEndpointScheme) {
+		input = strings.TrimPrefix(input, proxyEndpointScheme)
+	}
+	serverURL, err := url.Parse(proxyEndpointScheme + input)
+	if err != nil {
+		return nil, err
+	}
+	if serverURL.Hostname() == ecrPublicName {
+		return &Registry{
+			Service: ServiceECRPublic,
+		}, nil
 	}
-	matches := ecrPattern.FindStringSubmatch(serverURL)
+	matches := ecrPattern.FindStringSubmatch(serverURL.Hostname())
 	if len(matches) == 0 {
 		return nil, fmt.Errorf(programName + " can only be used with Amazon Elastic Container Registry.")
 	} else if len(matches) < 3 {
-		return nil, fmt.Errorf(serverURL + "is not a valid repository URI for Amazon Elastic Container Registry.")
+		return nil, fmt.Errorf(input + "is not a valid repository URI for Amazon Elastic Container Registry.")
```

Comment:
It might be better to have ` ` before "is not a valid ...". The current implementation results "foobaris not a valid ..."

Alternatively you can do "%q is not a valid ..."

Reply:
Done